snapshot-labs/pineapple
# fix: fix error not caught by try/catch

## 🧿 Current issues / What's wrong ?

When uploading something other than an image with `/upload`, it fails with an exception from sharp.

The try/catch located outside the async anonymous function is not catching the exception, making the process crash and exit (it get reboot automatically by digitalocean though). This is the cause behind the occasional status 504 detected by our monitoring system.

## 💊 Fixes / Solution

Rescue the exception, and handled it properly, to avoid crashing the whole process

Fix #141 
Fix PINEAPPLE-3

## 🚧 Changes

- Move the try/catch inside the anonymous async function, so all errors are caught correctly.
- `upload` exception are not handled by the try/catch anymore, but this function should not raise any exceptions, as all errors are returned by its 3rd argument (`upload(req, res, (error) => {)`)
- All user induced errors return a 40x error code, instead of 500

## 🛠️ Tests

E2E have been added to cover all cases regarding `/upload`.

To test the main fix manually, 
- use postman to upload something which is not image

Before 
- it crash the process

After
- it should return a json-rpc error, with 40x status code
- it should not crash the process
